File: src/lib/rename.ts
import type { Paper } from '@/types/paper';

const INVALID_FILENAME = /[<>:"/\\|?*]/g;

function cleanSegment(value: string): string {
  return value.replace(INVALID_FILENAME, ' ').replace(/\s+/g, ' ').trim();
}

export function renameSuggestion(paper: Paper): string {
  const firstAuthor = paper.authors[0] ?? 'UnknownAuthor';
  const title = paper.title || paper.file_name.replace(/\.pdf$/i, '');
  const year = paper.year ? `${paper.year}` : 'UnknownYear';
  const suggestion = `${year} - ${firstAuthor} - ${title}.pdf`;
  return cleanSegment(suggestion);
}

File: src-tauri/src/services/rename.rs
use sanitize_filename::sanitize;

use crate::models::Paper;

pub fn suggestion(paper: &Paper) -> String {
    let year = paper
        .year
        .map(|v| v.to_string())
        .unwrap_or_else(|| "UnknownYear".to_string());
    let first_author = paper
        .authors
        .first()
        .cloned()
        .unwrap_or_else(|| "UnknownAuthor".to_string());
    let title = if paper.title.is_empty() {
        paper.file_name.trim_end_matches(".pdf").to_string()
    } else {
        paper.title.clone()
    };
    let raw = format!("{} - {} - {}.pdf", year, first_author, title);
    sanitize(raw).replace('_', " ")
}
